File: backend/src/routes/account.ts
import type { FastifyPluginAsync } from "fastify";
import { and, eq } from "drizzle-orm";
import { db } from "../db/index.js";
import { usageEvents, userMemes, users } from "../db/schema.js";
import { deleteStoredImage } from "../services/image-downloader.js";
import { requireInstallUserId } from "./identity.js";

export const accountRoutes: FastifyPluginAsync = async (app) => {
  app.get("/account/export", async (request, reply) => {
    const userId = await requireInstallUserId(request, reply);
    if (!userId) return;

    const [library, usage] = await Promise.all([
      db
        .select()
        .from(userMemes)
        .where(eq(userMemes.userId, userId)),
      db
        .select()
        .from(usageEvents)
        .where(eq(usageEvents.userId, userId)),
    ]);

    return {
      install_id: userId,
      exported_at: new Date().toISOString(),
      library,
      usage_events: usage,
    };
  });

  app.delete("/account", async (request, reply) => {
    const userId = await requireInstallUserId(request, reply);
    if (!userId) return;

    const savedMemes = await db
      .select({ id: userMemes.id, filePath: userMemes.filePath })
      .from(userMemes)
      .where(eq(userMemes.userId, userId));

    let deletedFiles = 0;
    for (const meme of savedMemes) {
      try {
        if (await deleteStoredImage(meme.filePath)) deletedFiles++;
      } catch (err) {
        request.log.warn({ err, memeId: meme.id }, "Failed to delete stored meme file");
      }
    }

    const deletedUsage = await db
      .delete(usageEvents)
      .where(eq(usageEvents.userId, userId))
      .returning({ id: usageEvents.id });

    const deletedMemes = await db
      .delete(userMemes)
      .where(eq(userMemes.userId, userId))
      .returning({ id: userMemes.id });

    const deletedUsers = await db
      .delete(users)
      .where(and(eq(users.id, userId), eq(users.email, `install-${userId}@anonymous.memedrop.local`)))
      .returning({ id: users.id });

    return {
      deleted: true,
      install_id: userId,
      deleted_library_items: deletedMemes.length,
      deleted_usage_events: deletedUsage.length,
      deleted_files: deletedFiles,
      deleted_account: deletedUsers.length > 0,
    };
  });
};

File: apps/api/src/memedrop_api/api/account.py
from __future__ import annotations

import logging
from datetime import UTC, datetime

from fastapi import APIRouter, Request

from memedrop_api.identity import RequiredUserId
from memedrop_api.repositories import BackendStore

LOGGER = logging.getLogger("memedrop.account")
router = APIRouter(prefix="/api/v1", tags=["account"])


@router.get("/account/export")
async def export_account(
    request: Request,
    user_id: RequiredUserId,
) -> dict[str, object]:
    store: BackendStore = request.app.state.store
    library, usage = await store.export_account(user_id)
    return {
        "install_id": str(user_id),
        "exported_at": datetime.now(UTC).isoformat(),
        "library": library,
        "usage_events": usage,
    }


@router.delete("/account")
async def delete_account(
    request: Request,
    user_id: RequiredUserId,
) -> dict[str, object]:
    store: BackendStore = request.app.state.store
    saved_memes, deleted_memes, deleted_usage, deleted_user = await store.delete_account(user_id)
    deleted_files = 0
    for meme in saved_memes:
        try:
            deleted = await request.app.state.delete_stored_image(
                meme["filePath"], request.app.state.settings.meme_storage_path
            )
            deleted_files += int(deleted)
        except OSError:
            LOGGER.warning("Failed to delete stored meme", extra={"meme_id": meme["id"]})
    return {
        "deleted": True,
        "install_id": str(user_id),
        "deleted_library_items": deleted_memes,
        "deleted_usage_events": deleted_usage,
        "deleted_files": deleted_files,
        "deleted_account": deleted_user,
    }
